### PR TITLE
 feat(core): add config for update_mandate_flow 

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
 add config for update_mandate_flow and allow only those payment_methods that support update while Listing PaymentMethods For Merchants

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
- Create an MA and a MCA for Cybersource 
> Create an update_,mandate payment using cyber source it'll be successful
- Create  an MA and a MCA forStripe 
>  Create an update_,mandate payment using cyber source it should throw a 5xx

- Now ListPaymentMethodsForMerchnat in an Update Flow, we only list cards

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payment_methods/cards.rs [hunk 1]
```diff
@@ -2113,6 +2119,18 @@ pub async fn filter_payment_methods(
...
     Ok(())
 }
+fn filter_pm_based_on_supported_payments_for_update_mandate(
+    supported_payment_methods_for_mandate: &settings::SupportedPaymentMethodsForMandate,
+    payment_method_type: &api_enums::PaymentMethodType,
+    connector: api_enums::Connector,
+) -> bool {
+    supported_payment_methods_for_mandate
+        .0
+        .get(&api_enums::PaymentMethod::Card)
```
Comment (vspecky) [inline crates/router/src/core/payment_methods/cards.rs:2129]: Don't hardcode this to `Card`. We're using the Payment Method Type from the payment method object so the Payment Method should come from there too.

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -3889,3 +3892,12 @@ pub fn validate_session_expiry(session_expiry: u32) -> Result<(), errors::ApiErr
         Ok(())
     }
 }
+
+pub fn update_supported_connectors(
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:3896]: Nit: Is this function really required? Cuz it's too simple/small. If it's only being used at one place, remove it.

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs (hunk 1 above)
Comment (Aprabhat19) [reply crates/router/src/core/payment_methods/cards.rs:2129]: So should we construct a map that is specific to update_card, which has PaymentMethod with its respective PaymentMethodType and the connector that it supports ?

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -1011,6 +1011,23 @@ impl PaymentMethodData {
             self.to_owned()
         }
     }
+    pub fn get_payment_method_for_payment_method_data(&self) -> Option<api_enums::PaymentMethod> {
```
Comment (vspecky) [inline crates/api_models/src/payments.rs:1014]:
```suggestion
    pub fn get_payment_method(&self) -> Option<api_enums::PaymentMethod> {
```

File: crates/router/src/core/payment_methods/cards.rs [hunk 4]
```diff
@@ -2113,6 +2122,19 @@ pub async fn filter_payment_methods(
     }
     Ok(())
 }
+pub fn filter_pm_based_on_supported_payments_for_update_mandate(
```
Comment (vspecky) [inline crates/router/src/core/payment_methods/cards.rs:2125]:
```suggestion
pub fn filter_pm_based_on_update_mandate_support_for_connector(
```

File: crates/router/src/core/payments/flows/setup_mandate_flow.rs [hunk 5]
```diff
@@ -303,6 +254,130 @@ impl types::SetupMandateRouterData {
...
+        maybe_customer: &Option<domain::Customer>,
+    ) -> RouterResult<Self> {
+        let payment_method_type = self.request.payment_method_type;
+
+        let payment_method = self
+            .request
+            .payment_method_data
+            .get_payment_method_for_payment_method_data();
+
+        if payment_method_type.map_or(false, |pmt| {
```
Comment (vspecky) [inline crates/router/src/core/payments/flows/setup_mandate_flow.rs:277]: Put the condition result in a variable and use the variable in the if statement.

Comment (vspecky) [reply crates/router/src/core/payments/flows/setup_mandate_flow.rs:277]: Also
```suggestion
        payment_method.zip(payment_method_type).map_or(false, |(pm, pmt)| { ... })
```
is a better way to do it.

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

File: config/development.toml
Comment (pixincreate) [inline config/development.toml:1]: Similar to this, you can the configs for integration_test, sandbox and production toml files [here](https://github.com/juspay/hyperswitch/tree/main/config/deployments)

**Review by pixincreate — CHANGES_REQUESTED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

File: crates/router/src/core/payments/flows/setup_mandate_flow.rs [hunk 6]
```diff
@@ -303,6 +254,128 @@ impl types::SetupMandateRouterData {
...
+                }
+                Ok(_) => Err(errors::ApiErrorResponse::InternalServerError)
+                    .into_report()
+                    .attach_printable("Unexpected response received")?,
+                Err(_) => Ok(resp),
+            }
+        } else {
+            Err(errors::ApiErrorResponse::InternalServerError)
+                .into_report()
+                .attach_printable("Update Mandate Flow not implemented for the connector")?
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/flows/setup_mandate_flow.rs:376]: is it possible to add connector name here?

**Review by jagan-jaya — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

File: config/config.example.toml [hunk 7]
```diff
@@ -391,6 +391,10 @@ bank_redirect.ideal = {connector_list = "stripe,adyen,globalpay"} # Mandate supp
 bank_redirect.sofort = {connector_list = "stripe,adyen,globalpay"}
 wallet.apple_pay = { connector_list = "stripe,adyen,cybersource,noon" }
 
+[mandates.update_mandate_supported]
+card.credit ={connector_list ="cybersource"}  # Update Mandate supported payment method type and connector for card
```
Comment (pixincreate) [inline config/config.example.toml:395]: Nit (everywhere): 
```suggestion
card.credit = { connector_list = "cybersource" }  # Update Mandate supported payment method type and connector for card 
```

**Review by Narayanbhat166 — APPROVED** [review]

**Review by pixincreate — APPROVED** [review]

File: loadtest/config/development.toml [hunk 8]
```diff
@@ -252,6 +252,13 @@ bank_debit.sepa = { connector_list = "gocardless"}
 bank_redirect.ideal = {connector_list = "stripe,adyen,globalpay"}
 bank_redirect.sofort = {connector_list = "stripe,adyen,globalpay"}
 
+[mandates.update_mandate_supported]
+card.credit ={connector_list ="cybersource"}
+card.debit = {connector_list ="cybersource"}
+
+[mandates.update_mandate_supported]
+connector_list = "cybersource"
```
Comment (SanchithHegde) [inline loadtest/config/development.toml:260]: Possibly left this duplicate by mistake?

**Pull request merged**